Feature/return missing fields

albertshau opened the pull request.

albertshau committed:
COOPR-483 changing cluster create and delete requests to return
which required provider fields are missing if they are missing.

As part of this, removed provider fields from all requests except
creates and deletes since other operations do not require
communicating with providers. Also refactored cluster service a
bit to have a separate method for adding and validating user
given provider fields instead of baking it into other methods.
This was done in order to allow returning the missing fields.

albertshau committed:
updating docs to reflect api changes.

poornachandra commented on coopr-docs/docs/source/rest/clusters.rst:
```diff
@@ -105,6 +112,32 @@ Example
+                   "label": "SSH key",
+                   "override": false,
+                   "sensitive": true,
+                   "tip": "Your ssh key",
```
Missing "type" in "ssh_key".

poornachandra commented on coopr-docs/docs/source/rest/clusters.rst:
```diff
@@ -328,6 +375,30 @@ Example
+                   "label": "SSH key",
+                   "override": false,
+                   "sensitive": true,
+                   "tip": "Your ssh key",
```
Missing "type" here again.

poornachandra commented on coopr-server/src/main/java/co/cask/coopr/http/handler/ClusterHandler.java:
```diff
@@ -287,10 +289,18 @@ public void createCluster(HttpRequest request, HttpResponder responder) {
-      JsonObject response = new JsonObject();
-      response.addProperty("id", id);
-      responder.sendJson(HttpResponseStatus.OK, response);
+      Cluster cluster = clusterService.createEmptyCluster(clusterCreateRequest, account);
```
createEmptyCluster is a bit confusing since the cluster object will be eventually used to create a real cluster.

poornachandra commented on coopr-server/src/main/java/co/cask/coopr/http/handler/ClusterHandler.java:
```diff
@@ -287,10 +289,18 @@ public void createCluster(HttpRequest request, HttpResponder responder) {
+      if (!missingFields.isEmpty()) {
+        // nested structure means we have to provide a type
+        responder.sendJson(HttpResponseStatus.BAD_REQUEST, ImmutableMap.of("missingFields", missingFields),
+                           new TypeToken<Map<String, List<Map<String, FieldSchema>>>>() {}.getType());
```
new TypeToken<Map<String, List<Map<String, FieldSchema>>>>() {}.getType() can be made into a static final Type. Also, generally for serializing we may not need to provide Type as Guava can figure out the type through reflection.

albertshau committed:
adding provider fields back to all cluster operation requests
since things like ssh keys may be in the provider and would be
required by provisioner to ssh to nodes and perform actions.

Restructuring ClusterService to look more like it did before,
and throwing a specific exception when fields are missing instead
of trying to return them directly. This because most of what
happens in each method has to stay there for transactionality and
cannot be broken into different phases.

albertshau committed:
add copyright

poornachandra commented on coopr-docs/docs/source/rest/clusters.rst:
```diff
@@ -328,6 +376,31 @@ Example
+                   "override": false,
+                   "sensitive": true,
+                   "tip": "Your ssh key",
+		   "type": "file"
```
alignment issue.

albertshau committed:
undo unintended callback changes.

albertshau committed:
fix doc issues and give more details about sensitive fields.

poornachandra commented on coopr-docs/docs/source/rest/clusters.rst:
```diff
@@ -77,7 +77,14 @@ Required Parameters
-The server will respond with the id of the cluster added.
+The server will respond with the id of the cluster added. If there are required provider fields missing from
+the request, the server will respond with a JSON Object describing the missing fields. The response will
+contain a ``missingFields`` key which will be an array of JSON Objects containing the missing field as the
```
This description can refer to the section on :ref:`Provider Fields <cluster-provider-fields>`

poornachandra commented on coopr-docs/docs/source/rest/clusters.rst:
```diff
@@ -299,7 +306,13 @@ To delete a cluster, make a DELETE HTTP request to URI:
-This resource request represents an individual cluster for deletion.
+This resource request represents an individual cluster for deletion. If the provider used to create the cluster
+uses sensitive fields, it is possible the sensitive fields were lost at some point and need to be given again.
+To do this, the request must contain a JSON Object as its body, with ``providerFields`` as a key, and a JSON
```
This description can refer to the section on :ref:`Provider Fields <cluster-provider-fields>`

poornachandra commented:
Minor comment, otherwise LGTM :+1:

albertshau committed:
reference the section in the doc instead of repeating information.
